### PR TITLE
test(gateway): integration gateway test with  authn

### PR DESCRIPTION
**Pull request by lafirest**

### CODE REVIEW

File: apps/emqx_gateway/test/emqx_gateway_authn_http_SUITE.erl [hunk 1]
```diff
@@ -0,0 +1,379 @@
...
+    Config.
+
+end_per_testcase(_Case, Config) ->
+    Config.
+
+%%------------------------------------------------------------------------------
+%% Tests
+%%------------------------------------------------------------------------------
+t_authn(_) ->
+    test_gateway_with_auths(?CASES, ?AUTHNS).
```
Comment (HJianBo) [inline apps/emqx_gateway/test/emqx_gateway_authn_http_SUITE.erl:103]: Can we put the logic for starting/stopping the gateway and deps resources into another module? In that way, we only need to focus on how to organize the tests in `*_SUITE.erl`. In the _SUITE.erl, It is preferred:

```erlang

groups() ->
  emqx_gateway_ct:init_per_groups(matrix()).

matrix() ->
  %% Gateways :: [lwm2m, mqttsn, stomp, ...] | all
  %% Authns :: [auth_http, built_in_database, ...]
  %% Authzs :: [auth_http, built_in_database, ...]
  {GatewayNames, AuthnNames, AuthzNames}.

init_per_group(GrpName, _Conf) ->
  emqx_gateway_ct:init_per_group(GrpName, fun prepare/1).

end_per_group(GrpName, _Conf) ->
   emqx_gateway_ct:end_per_group(GrpName, fun cleanup/1).

prepare({GatewayName, AuthnName, AuthzName) ->
    _ = start_authn_server(AuthnName),
    _ = start_authz_server(AuthzName),
   emqx_gateway_ct:merge_confs(
    gateway_confs(GatewayName),
    authn_confs(AuthnName),
    authz_confs(AuthzName)).

cleanup({GatewayName, AuthnName, AuthzName) ->
  _ = clean_authn_resources(AuthnName),
  _ = clean_authz_resources(AuthzName),
  ok.

%% cases

t_connect_successfully(Conf) ->
    GwName = emqx_gateway_ct:gateway(Conf),
   emqx_gateway_ct:with_client(GwName, fun(C) ->
       ok
   end).

...

```

File: apps/emqx_gateway/test/emqx_gateway_authn_http_SUITE.erl [hunk 2]
```diff
@@ -0,0 +1,379 @@
...
+
+case_coap() ->
+    Login = fun(URI, Checker) ->
+        Action = fun(Channel) ->
+            Req = emqx_coap_SUITE:make_req(post),
+            Checker(emqx_coap_SUITE:do_request(Channel, URI, Req))
+        end,
+        emqx_coap_SUITE:do(Action)
+    end,
+    Prefix = emqx_coap_SUITE:mqtt_prefix(),
```
Comment (HJianBo) [inline apps/emqx_gateway/test/emqx_gateway_authn_http_SUITE.erl:117]: These calls seem to be poorly maintained.
Can we encapsulate all these client-side-related operations in a separate module? i.e: `emqx_gateway_ct_coap_client.erl`

Comment (lafirest) [reply apps/emqx_gateway/test/emqx_gateway_authn_http_SUITE.erl]: I think we can't make everything better in one PR, the optimisation of clients maybe be another task

File: apps/emqx_authn/src/simple_authn/emqx_authn_http.erl [hunk 3]
```diff
@@ -383,3 +384,14 @@ to_bin(L) when is_list(L) ->
...
+ensure_header_name_type(Headers) ->
+    Fun = fun
+        (Key, _Val, Acc) when is_binary(Key) ->
+            Acc;
+        (Key, Val, Acc) when is_atom(Key) ->
+            Acc2 = maps:remove(Key, Acc),
+            BinKey = erlang:atom_to_binary(Key),
+            Acc2#{BinKey => Val}
+    end,
+    maps:fold(Fun, Headers, Headers).
```
Comment (HJianBo) [inline apps/emqx_authn/src/simple_authn/emqx_authn_http.erl:397]: nit: `maps:map/2` is better for me

**Review by HJianBo — APPROVED** [review]

**Pull request merged**